src/compiler/debug.ts

```diff
@@ -301,19 +301,19 @@ namespace ts {
                 return members.length > 0 && members[0][0] === 0 ? members[0][1] : "0";
             }
             if (isFlags) {
-                let result = "";
+                const result: string[] = [];

```

Comment:
I'm not sure if it's still the case, but in earlier v8 versions if you avoid allocating the array until you have the first element the performance is improved because it doesn't have to re-type the array.

```js
const ar1 = []; // typed internally as a JS_ARRAY_TYPE of PACKED_SMI_ELEMENTS
ar1.push("foo"); // deoptimizes, re-types, now it's internally typed as an array of strings

// vs

const ar2 = ["foo"]; // typed internally as an array of strings
```

As a result, we have a lot of code that does either of the following:

```ts
let array: Foo[] | undefined;
...
if (array) {
  array.push(foo);
}
else {
  array = [foo];
}

// or
...
array = append(array, foo); // same as above, but in an inlinable function
```

That said, since this code isn't performance critical, it probably wouldn't help much.

Reply:
That's an interesting idea; I know that this string addition itself was definitely bad, but I had never (when fixing this kind of code after pprof showing huge allocations) thought about creating a single array with one element to get v8 to do the right thing; the performance boost from just doing `join` was the main significant item.

Reply:
IIRC, this was a performance optimization employed by @vladima back in the day. It still seems to be effective from what I can tell.